Mathlib/Data/Rel.lean

```diff
@@ -4,370 +4,312 @@ Released under Apache 2.0 license as described in the file LICENSE.
-import Mathlib.Tactic.AdaptationNote
+import Mathlib.Tactic.MinImports
 
 /-!
-# Relations
+# Relations as sets of pairs
 
-This file defines bundled relations. A relation between `α` and `β` is a function `α → β → Prop`.
-Relations are also known as set-valued functions, or partial multifunctions.
+This file provides API to regard relations between `α` and `β`  as sets of pairs `Set (α × β)`.
+
+This is in particular useful in the study of uniform spaces, which are topological spaces equipped
+with a *uniformity*, namely a filter of pairs `α × α` whose elements can be viewed as "proximity"
+relations.
 
 ## Main declarations
 
-* `Rel α β`: Relation between `α` and `β`.
-* `Rel.inv`: `r.inv` is the `Rel β α` obtained by swapping the arguments of `r`.
-* `Rel.dom`: Domain of a relation. `x ∈ r.dom` iff there exists `y` such that `r x y`.
-* `Rel.codom`: Codomain, aka range, of a relation. `y ∈ r.codom` iff there exists `x` such that
-  `r x y`.
-* `Rel.comp`: Relation composition. Note that the arguments order follows the `CategoryTheory/`
-  one, so `r.comp s x z ↔ ∃ y, r x y ∧ s y z`.
-* `Rel.image`: Image of a set under a relation. `r.image s` is the set of `f x` over all `x ∈ s`.
-* `Rel.preimage`: Preimage of a set under a relation. Note that `r.preimage = r.inv.image`.
-* `Rel.core`: Core of a set. For `s : Set β`, `r.core s` is the set of `x : α` such that all `y`
-  related to `x` are in `s`.
-* `Rel.restrict_domain`: Domain-restriction of a relation to a subtype.
+* `Rel α β`: Type of relations between `α` and `β`.
+* `Rel.inv`: Turn `R : Rel α β` into `R.inv : Rel β α` by swapping the arguments.
+* `Rel.dom`: Domain of a relation. `a ∈ R.dom` iff there exists `b` such that `a ~[R] b`.
+* `Rel.cod`: Codomain of a relation. `b ∈ R.cod` iff there exists `a` such that `a ~[R] b`.
+* `Rel.id`: The identity relation `Rel α α`.
+* `Rel.comp`: Relation composition. Note that the arguments order follows the category theory
+  convention, namely `(R ○ S) a c ↔ ∃ b, a ~[R] b ∧ b ~[S] z`.
+* `Rel.image`: Image of a set under a relation. `b ∈ image R s` iff there exists `a ∈ s`
+  such that `a ~[R] b`.
+* `Rel.preimage`: Image of a set under a relation. `a ∈ image R t` iff there exists `b ∈ t`
+  such that `a ~[R] b`.
+* `Rel.core`: Core of a set. For `t : Set β`, `a ∈ R.core t` iff all `b` related to `a` are in `t`.
+* `Rel.restrictDomain`: Domain-restriction of a relation to a subtype.
 * `Function.graph`: Graph of a function as a relation.
+-/
 
-## TODO
+variable {α β γ δ : Type*}
 
-The `Rel.comp` function uses the notation `r • s`, rather than the more common `r ∘ s` for things
-named `comp`. This is because the latter is already used for function composition, and causes a
-clash. A better notation should be found, perhaps a variant of `r ∘r s` or `r; s`.
+/-- A relation on `α` and `β`, aka a set-valued function, aka a partial multifunction.
 
--/
+We represent them as sets due to how relations are used in the context of uniform spaces. -/
```

Comment:
I think this should have rather a longer explanation, perhaps in the module docstring. Perhaps some of your PR description can be repurposed.

What you have right now is not enough to persuade someone reading the source against changing to the `_ -> _ -> Prop` spelling again.

Reply:
What about now?